source-test/File.ts

```diff
@@ -1,661 +1,1617 @@
-      should.not.exist(file.stat);
+    it('defaults stat to null', function(done) {
+      const file = new Vinyl();
+      expect(file.stat).toNotExist();
+      expect(file.stat).toEqual(null);
       done();
     });
 
-    it('should default contents to null', done => {
-      const file = new File();
-      should.not.exist(file.contents);
+    it('defaults contents to null', function(done) {
+      const file = new Vinyl();
+      expect(file.contents).toNotExist();
+      expect(file.contents).toEqual(null);
       done();
     });
+  });
 
-    it('should set base to given value', done => {
-      const val = '/';
-      const file = new File({ base: val });
-      file.basename.should.equal(val);
+  describe('constructor()', function() {
+
+    it('sets base', function(done) {
+      const val = path.normalize('/');
+      const file = new Vinyl({ base: val });
+      expect(file.base).toEqual(val);
       done();
     });
 
-    it('should set cwd to given value', done => {
-      const val = '/';
-      const file = new File({ cwd: val });
-      file.cwd.should.equal(val);
+    it('sets cwd', function(done) {
+      const val = path.normalize('/');
+      const file = new Vinyl({ cwd: val });
+      expect(file.cwd).toEqual(val);
       done();
     });
 
-    it('should set path to given value', done => {
-      const val = '/test.coffee';
-      const file = new File({ path: val });
-      file.path.should.equal(val);
+    it('sets path (and history)', function(done) {
+      const val = path.normalize('/test.coffee');
+      const file = new Vinyl({ path: val });
+      expect(file.path).toEqual(val);
+      expect(file.history).toEqual([val]);
       done();
     });
 
-    it('should set stat to given value', done => {
-      const val = {};
-      const file = new File(<fs.Stats><any>{ stat: val });
-      file.stat.should.equal(val);
+    it('sets history (and path)', function(done) {
+      const val = path.normalize('/test.coffee');
+      const file = new Vinyl({ history: [val] });
+      expect(file.path).toEqual(val);
+      expect(file.history).toEqual([val]);
       done();
     });
 
-    it('should set contents to given value', done => {
+    it('sets stat', function(done) {
+      const val: any = {};
+      const file = new Vinyl({ stat: val });
+      expect(file.stat).toEqual(val);
+      done();
+    });
+
+    it('sets contents', function(done) {
       const val = new Buffer('test');
-      const file = new File({ contents: val });
-      file.contents.should.equal(val);
+      const file = new Vinyl({ contents: val });
+      expect(file.contents).toEqual(val);
       done();
     });
 
-    it('should default basename to cwd', done => {
-      const cwd = '/';
-      const file = new File({ cwd: cwd });
-      file.basename.should.equal(cwd);
+    it('sets custom properties', function(done) {
+      const sourceMap = {};
+      const file = <Vinyl & {sourceMap: {}}> new Vinyl(<any> { sourceMap: sourceMap });
```

Comment:
I was wondering if it was possible to tell typescript that the class implements its options. Here is a solution I tried but it did not seem to work:

```typescript
interface FileOptions {
  src: string;
}
// This does not work
class File<O extends {}> implements O {
  constructor (options?: FileOptions & O);
}

// This does not work either
class File<O extends {}> {
  constructor (options?: FileOptions & O): this & O;
}
```

The best I could come up with is the solution I used in the tests where the user augments the object:

```typescript
interface Custom {
  custom: number;
}

const file = <Vinyl & Custom> new Vinyl(<FileOptions & Custom> {src: "...", custom: 10});
// It requires more work from the user, but why not. The problem is `clone`:
const file2 = file.clone();  // file2 is typed as Vinyl, not Vinyl & Custom (you have to cast it manually)
```

Do you have some ideas ?

(Of course, I'll add the string property accessor on options, but how to express the dependency between the options and the properties of the object ?)